### PR TITLE
[FLINK-27675] Improve manual savepoint tracking

### PR DESCRIPTION
**Pull request by morhidi**
Cancelling savepoint operation on application failures
Cancelling savepoint operation on savepoint fetching failures
Generating events for failed savepoint operations

### CODE REVIEW

**Comment by morhidi** [comment]
cc @wangyang0918 @SteNicholas

**Review by wangyang0918 — APPROVED** [review]
+1

**Comment by wangyang0918** [comment]
I am doing a manual verification in my minikube.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/SavepointObserver.java [hunk 1]
```diff
@@ -106,36 +117,36 @@ private Optional<String> observeTriggeredSavepointProgress(
...
+                String errorMsg =
+                        "Savepoint triggering timed out after "
+                                + configManager
+                                        .getOperatorConfiguration()
+                                        .getSavepointTriggerGracePeriod();
                 currentSavepointInfo.resetTrigger();
+                LOG.error(errorMsg);
                 return Optional.of(errorMsg);
+            } else {
+                LOG.info("Savepoint operation not running yet, waiting within grace period...");
```
Comment (wangyang0918) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/SavepointObserver.java:137]: not running yet -> not finished yet? The savepoint should already be triggered successfully.

Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/SavepointObserver.java]: fixed

**Comment by morhidi** [comment]
> I am doing a manual verification in my minikube.

thanks!

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java [hunk 2]
```diff
@@ -98,6 +98,10 @@ public void observe(FlinkDeployment flinkApp, Context context) {
             }
         }
 
+        if (!ReconciliationUtils.isJobRunning(flinkApp.getStatus())) {
```
Comment (wangyang0918) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java:101]: We might also need to create an event when we canceling the savepoint operation. At least, we need to add a log when doing a concrete reset(when the `triggerId` is not empty).

**Review by wangyang0918 — COMMENTED** [review]
Works well after manual verification. After this PR, the savepoint failure now is more easily to be found.

```
Events:
  Type     Reason          Age   From      Message
  ----     ------          ----  ----      -------
  Warning  SavepointError  33m   Operator  Savepoint failed for savepointTriggerNonce: 2
  Warning  SavepointError  17m   Operator  Savepoint failed for savepointTriggerNonce: 3
```

The savepoint process still could be improved in FLINK-27257 by supporting retrying at some given exception.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java (hunk 2 above)
Comment (morhidi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java:101]: Added, PTAL

**Comment by wangyang0918** [comment]
Will merge this PR after CI pass.

**Pull request merged**